File: libraries/python/src/vsync_s3_client/config_blob.py
"""`VSYNC_CONFIG` bootstrap blob decoder.

Wire format (v0.12 §2.1):
    ``vsync-cfg-v1:<base64url-no-pad(gzip(JSON))>``

The magic prefix is also the schema-version handle:
- absent / non-`vsync-cfg-v1:` → ConfigMissingError
- present, base64url body decodes to non-gzip → BundleCorruptError
- present, gzip ok, JSON inner `v != 1` → ConfigUnsupportedVersionError
- standard base64 (`+`, `/`, `=`) MUST be rejected → ConfigUnsupportedVersionError
  (the receiver can't tell whether you intended url-safe and corrupted it,
  or intended standard and shipped the wrong alphabet; either way they're
  holding a blob outside the contract)
"""

from __future__ import annotations

import base64
import gzip
import json
from dataclasses import dataclass
from typing import Union

from .exceptions import (
    BundleCorruptError,
    ConfigMissingError,
    ConfigUnsupportedVersionError,
)

BLOB_MAGIC = "vsync-cfg-v1:"
SUPPORTED_INNER_V = 1
# RFC 4648 §5 alphabet uses '-' and '_' instead of '+' and '/'.
_DISALLOWED_BASE64_CHARS = frozenset("+/=")


@dataclass(frozen=True)
class VsyncConfig:
    """Decoded inner JSON of the `VSYNC_CONFIG` blob.

    Field naming is snake_case on the Python side (idiomatic) while the
    wire JSON keys remain camelCase per v0.12 §2.1. Mapping happens in
    `decode_config_blob`.

    `salt` + `iterations` carry the PBKDF2 parameters needed to derive
    the AES-256 key from `VSYNC_PASSPHRASE`. They land in the blob from
    `vsync runtime-token`; both are required (v0.12 §2.1, post-revision).
    """

    v: int
    endpoint: str
    region: str
    bucket: str
    access_key_id: str
    secret_access_key: str
    prefix: str
    env: str
    salt: str
    iterations: int

    def __repr__(self) -> str:  # redaction — never log credentials
        return (
            f"VsyncConfig(v={self.v}, endpoint={self.endpoint!r}, "
            f"region={self.region!r}, bucket={self.bucket!r}, "
            f"env={self.env!r}, prefix={self.prefix!r}, "
            f"iterations={self.iterations}, "
            f"accessKeyId=<redacted>, secretAccessKey=<redacted>, "
            f"salt=<redacted>)"
        )


def _as_bytes(blob: Union[bytes, bytearray, memoryview, str]) -> bytes:
    if isinstance(blob, str):
        return blob.encode("ascii")
    if isinstance(blob, (bytes, bytearray, memoryview)):
        return bytes(blob)
    raise ConfigMissingError(
        f"VSYNC_CONFIG: expected bytes or str, got {type(blob).__name__}"
    )


def decode_config_blob(
    blob: Union[bytes, bytearray, memoryview, str],
) -> VsyncConfig:
    """Decode a `VSYNC_CONFIG` blob → typed config.

    Raises:
        ConfigMissingError: magic prefix absent or wrong (incl. raw JSON,
            wrong-version magic, empty input).
        ConfigUnsupportedVersionError: inner JSON `v` > 1, or the body uses
            the standard base64 alphabet instead of base64url-no-pad.
        BundleCorruptError: magic + base64 ok but the bytes inside aren't
            gzip / aren't a JSON object.
    """
    data = _as_bytes(blob)
    if not data.startswith(BLOB_MAGIC.encode("ascii")):
        raise ConfigMissingError(
            "VSYNC_CONFIG: missing 'vsync-cfg-v1:' prefix — did you paste "
            "raw JSON, or are you holding a newer (v2+) blob?"
        )
    body = data[len(BLOB_MAGIC):]

    # Strict base64url rejection: any standard-alphabet character or
    # explicit padding is the operator hand-rolling with `base64` instead
    # of `base64url`. Surface that loudly.
    for ch in _DISALLOWED_BASE64_CHARS:
        if ch.encode("ascii") in body:
            raise ConfigUnsupportedVersionError(
                "VSYNC_CONFIG: body must be base64url-no-pad per v0.12 §2.1; "
                f"found disallowed character {ch!r} (use '-' and '_' instead "
                "of '+' and '/'; drop padding '=')"
            )

    # base64url without padding — re-add padding for the decoder.
    pad = b"=" * ((-len(body)) % 4)
    try:
        decoded = base64.urlsafe_b64decode(body + pad)
    except Exception as e:  # noqa: BLE001 - convert to taxonomy
        raise BundleCorruptError(
            f"VSYNC_CONFIG: base64url body failed to decode: {e}"
        ) from e

    try:
        raw_json = gzip.decompress(decoded)
    except Exception as e:  # noqa: BLE001 - covers OSError + EOFError
        raise BundleCorruptError(
            f"VSYNC_CONFIG: gzip decompress failed — body is not a gzip stream: {e}"
        ) from e

    try:
        obj = json.loads(raw_json)
    except json.JSONDecodeError as e:
        raise BundleCorruptError(
            f"VSYNC_CONFIG: inner JSON failed to parse: {e}"
        ) from e
    if not isinstance(obj, dict):
        raise BundleCorruptError(
            f"VSYNC_CONFIG: inner JSON must be an object, got {type(obj).__name__}"
        )

    v = obj.get("v")
    if v != SUPPORTED_INNER_V:
        raise ConfigUnsupportedVersionError(
            f"VSYNC_CONFIG: inner v={v!r}; this library understands v=1 only — "
            "upgrade vsync-s3-client"
        )

    # Required fields per v0.12 §2.1. Missing field → BundleCorrupt (the
    # operator's blob is internally inconsistent with the spec).
    try:
        iters = obj["iterations"]
        if not isinstance(iters, int) or iters <= 0:
            raise BundleCorruptError(
                f"VSYNC_CONFIG: iterations must be a positive int, got {iters!r}"
            )
        return VsyncConfig(
            v=int(obj["v"]),
            endpoint=str(obj["endpoint"]),
            region=str(obj["region"]),
            bucket=str(obj["bucket"]),
            access_key_id=str(obj["accessKeyId"]),
            secret_access_key=str(obj["secretAccessKey"]),
            prefix=str(obj["prefix"]),
            env=str(obj["env"]),
            salt=str(obj["salt"]),
            iterations=iters,
        )
    except KeyError as e:
        raise BundleCorruptError(
            f"VSYNC_CONFIG: inner JSON is missing required field {e.args[0]!r}"
        ) from e


__all__ = ["VsyncConfig", "decode_config_blob"]

File: libraries/java/src/main/java/io/github/muthuishere/vsync/s3client/config/ConfigBlob.java
package io.github.muthuishere.vsync.s3client.config;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.github.muthuishere.vsync.s3client.exceptions.BundleCorruptException;
import io.github.muthuishere.vsync.s3client.exceptions.ConfigMissingException;
import io.github.muthuishere.vsync.s3client.exceptions.ConfigUnsupportedVersionException;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.Base64;
import java.util.zip.GZIPInputStream;

/**
 * {@code VSYNC_CONFIG} bootstrap blob decoder.
 *
 * <p>Wire format (v0.12 §2.1):
 * <pre>vsync-cfg-v1:&lt;base64url-no-pad(gzip(JSON))&gt;</pre>
 *
 * <p>The magic prefix is also the schema-version handle:
 * <ul>
 *   <li>absent / non-{@code vsync-cfg-v1:} → {@link ConfigMissingException}</li>
 *   <li>present, base64url body decodes to non-gzip → {@link BundleCorruptException}</li>
 *   <li>present, gzip ok, JSON inner {@code v != 1} → {@link ConfigUnsupportedVersionException}</li>
 *   <li>standard-base64 characters in the body ({@code +}, {@code /}, {@code =}) →
 *       {@link ConfigUnsupportedVersionException} (operator hand-rolled with the wrong
 *       alphabet — refuse rather than silently translate)</li>
 * </ul>
 */
public final class ConfigBlob {

    private static final String BLOB_MAGIC = "vsync-cfg-v1:";
    private static final byte[] BLOB_MAGIC_BYTES = BLOB_MAGIC.getBytes(StandardCharsets.US_ASCII);
    private static final int SUPPORTED_INNER_V = 1;

    /**
     * Sanity floor on the salt's string length. The CLI emits 24-char base64url
     * ASCII; floor at 16 so a typo'd / truncated blob fails fast. The bytes fed
     * to PBKDF2 are these chars' UTF-8 encoding verbatim — NOT base64-decoded.
     */
    private static final int MIN_SALT_CHARS = 16;

    private static final ObjectMapper MAPPER = new ObjectMapper();

    private ConfigBlob() {
    }

    public static VsyncConfig decode(byte[] blob) {
        if (blob == null || blob.length < BLOB_MAGIC_BYTES.length
                || !startsWith(blob, BLOB_MAGIC_BYTES)) {
            throw new ConfigMissingException(
                    "VSYNC_CONFIG: missing 'vsync-cfg-v1:' prefix — "
                            + "did you paste raw JSON, or are you holding a newer (v2+) blob?");
        }
        byte[] body = new byte[blob.length - BLOB_MAGIC_BYTES.length];
        System.arraycopy(blob, BLOB_MAGIC_BYTES.length, body, 0, body.length);

        // Strict base64url rejection. Any '+', '/', '=' is the standard alphabet —
        // refuse loudly rather than silently re-encode (which would mask the
        // operator's mistake).
        for (byte b : body) {
            if (b == '+' || b == '/' || b == '=') {
                throw new ConfigUnsupportedVersionException(
                        "VSYNC_CONFIG: body must be base64url-no-pad per v0.12 §2.1; "
                                + "found disallowed character " + (char) b
                                + " (use '-' and '_' instead of '+' and '/'; drop padding '=')");
            }
        }

        byte[] decoded;
        try {
            decoded = Base64.getUrlDecoder().decode(body);
        } catch (IllegalArgumentException e) {
            throw new BundleCorruptException(
                    "VSYNC_CONFIG: base64url body failed to decode: " + e.getMessage(), e);
        }

        // Gzip magic sniff: catches the "valid base64 of junk bytes" case before
        // we burn cycles in GZIPInputStream's reader.
        if (decoded.length < 2 || decoded[0] != (byte) 0x1f || decoded[1] != (byte) 0x8b) {
            throw new BundleCorruptException(
                    "VSYNC_CONFIG: body is not a gzip stream (wrong magic bytes)");
        }

        byte[] rawJson;
        try (GZIPInputStream gz = new GZIPInputStream(new ByteArrayInputStream(decoded))) {
            rawJson = gz.readAllBytes();
        } catch (IOException e) {
            throw new BundleCorruptException(
                    "VSYNC_CONFIG: gzip decompress failed: " + e.getMessage(), e);
        }

        JsonNode obj;
        try {
            obj = MAPPER.readTree(rawJson);
        } catch (JsonProcessingException e) {
            throw new BundleCorruptException(
                    "VSYNC_CONFIG: inner JSON failed to parse: " + e.getMessage(), e);
        } catch (IOException e) {
            throw new BundleCorruptException(
                    "VSYNC_CONFIG: inner JSON read failed: " + e.getMessage(), e);
        }
        if (!obj.isObject()) {
            throw new BundleCorruptException(
                    "VSYNC_CONFIG: inner JSON must be an object, got " + obj.getNodeType());
        }

        // Probe `v` first so the corpus's negative-unknown-version vector
        // surfaces the right error before any field-level checks.
        JsonNode vNode = obj.get("v");
        if (vNode == null || !vNode.isInt() || vNode.intValue() != SUPPORTED_INNER_V) {
            throw new ConfigUnsupportedVersionException(
                    "VSYNC_CONFIG: inner v=" + (vNode == null ? "missing" : vNode.asText())
                            + "; this library understands v=1 only — upgrade vsync-s3-client");
        }

        try {
            int iterations = requireInt(obj, "iterations");
            if (iterations <= 0) {
                throw new BundleCorruptException(
                        "VSYNC_CONFIG: iterations must be > 0, got " + iterations);
            }
            String salt = requireString(obj, "salt");
            if (salt.length() < MIN_SALT_CHARS) {
                throw new ConfigUnsupportedVersionException(
                        "VSYNC_CONFIG: salt string is " + salt.length()
                                + " chars (< " + MIN_SALT_CHARS + " minimum)");
            }
            return new VsyncConfig(
                    SUPPORTED_INNER_V,
                    requireString(obj, "endpoint"),
                    requireString(obj, "region"),
                    requireString(obj, "bucket"),
                    requireString(obj, "accessKeyId"),
                    requireString(obj, "secretAccessKey"),
                    requireString(obj, "prefix"),
                    requireString(obj, "env"),
                    salt,
                    iterations);
        } catch (MissingFieldException e) {
            throw new BundleCorruptException(
                    "VSYNC_CONFIG: inner JSON is missing required field '" + e.field + "'", e);
        }
    }

    private static boolean startsWith(byte[] data, byte[] prefix) {
        if (data.length < prefix.length) {
            return false;
        }
        for (int i = 0; i < prefix.length; i++) {
            if (data[i] != prefix[i]) {
                return false;
            }
        }
        return true;
    }

    private static String requireString(JsonNode obj, String field) {
        JsonNode n = obj.get(field);
        if (n == null || n.isNull()) {
            throw new MissingFieldException(field);
        }
        return n.asText();
    }

    private static int requireInt(JsonNode obj, String field) {
        JsonNode n = obj.get(field);
        if (n == null || n.isNull()) {
            throw new MissingFieldException(field);
        }
        if (!n.canConvertToInt()) {
            throw new BundleCorruptException(
                    "VSYNC_CONFIG: " + field + " must be an int, got " + n.getNodeType());
        }
        return n.intValue();
    }

    private static final class MissingFieldException extends RuntimeException {
        final String field;

        MissingFieldException(String field) {
            super("missing field: " + field);
            this.field = field;
        }
    }
}
